compiler/src/vm/mod.rs

```diff
@@ -1,8 +1,854 @@
+pub struct FiberId(usize);
+
+#[derive(Clone)]
+enum FiberTree {
+    /// This tree is currently focused on running a single fiber.
+    Single(Single),
+
+    /// The fiber of this tree entered a `core.parallel` scope so that it's now
+    /// paused and waits for the parallel scope to end. Instead of the main
+    /// former single fiber, the tree now runs the closure passed to
+    /// `core.parallel` as well as any other spawned children.
+    Parallel(Parallel),
+
+    Try(Try),
+}
+
+/// Single fibers are the leafs of the fiber tree.
+#[derive(Clone)]
+struct Single {
+    fiber: Fiber,
+    parent: Option<FiberId>,
+}
+
+/// When a parallel section is entered, the fiber that started the section is
+/// paused. Instead, the children of the parallel section are run. Initially,
+/// there's only one child – the closure given to the parallel builtin function.
+/// Using the nursery parameter (a nursery can be thought of as a pointer to a
+/// parallel section), you can also spawn other fibers. In contrast to the first
+/// child, those children also have an explicit send port where the closure's
+/// result is sent to.
+#[derive(Clone)]
+struct Parallel {
+    paused_fiber: Single,
+    children: HashMap<FiberId, ChildKind>,
+    return_value: Option<Packet>, // will later contain the body's return value
+    nursery: ChannelId,
+}
+#[derive(Clone)]
+enum ChildKind {
+    InitialChild,
+    SpawnedChild(ChannelId),
+}
+
+#[derive(Clone)]
+struct Try {
+    paused_fiber: Single,
+    child: FiberId,
+}
+
+#[derive(Clone)]
+enum Channel {
+    Internal(InternalChannel),
+    External,
+    Nursery(FiberId),
+}
+#[derive(Clone, Debug)]
+struct InternalChannel {
+    buffer: ChannelBuf,
+    pending_sends: VecDeque<(Option<FiberId>, Packet)>,
+    pending_receives: VecDeque<Option<FiberId>>,
+}
+
+#[derive(Clone)]
+pub enum Operation {
+    Send {
+        performing_fiber: Option<FiberId>,
+        packet: Packet,
+    },
+    Receive {
+        performing_fiber: Option<FiberId>,
+    },
+    Drop,
+}
+
+#[derive(Clone, Debug)]
+pub enum Status {
+    CanRun,
+    WaitingForOperations,
+    Done,
+    Panicked { reason: String },
+}
+
+impl Vm {
+    pub fn new() -> Self {
+        Self {
+            channels: Default::default(),
+            fibers: HashMap::new(),
+            root_fiber: None,
+            external_operations: Default::default(),
+            channel_id_generator: IdGenerator::start_at(0),
+            fiber_id_generator: IdGenerator::start_at(0),
+        }
+    }
+
+    fn set_up_with_fiber(&mut self, fiber: Fiber) {
+        assert!(self.root_fiber.is_none(), "VM already set up");
+        let root_fiber_id = self.fiber_id_generator.generate();
+        self.fibers.insert(
+            root_fiber_id,
+            FiberTree::Single(Single {
+                fiber,
+                parent: None,
+            }),
+        );
+        self.root_fiber = Some(root_fiber_id);
+    }
+    pub fn set_up_for_running_closure(
```

Comment:
For the `main` function, we want to be able to call `create_channel` on the VM before calling `set_up_for_running_closure` with the environment that then contains the returned `ChannelId`s.

We could of course just pass an `IdGenerator<ChannelId>` into the constructors as well, would be nicer?

Reply:
Right, I saw that later on but forgot to remove this comment. I'm not sure an `IdGenerator<ChannelId>` would suffice since the VM has to know which channel(s) to create initially. I guess the current way suffices. If VMs are created in more places, we could think about a builder to encapsulate the setup in a type-safe way